### ISSUE TITLE
Turning off `strict-unsealed-patmat` doesn't work when unsealed type is nested

### ISSUE THREAD

**Issue by ghik** [scala/bug#12314]
## reproduction steps

Using Scala 2.13.4

Compile the following code with `strict-unsealed-patmat` disabled (i.e. `-Xlint:-strict-unsealed-patmat`):

```scala
final class Thing private()
object Thing {
  final val Foo = new Thing
  final val Bar = new Thing
}

object test extends App {
  val thingOpt: Option[Thing] = Some(Thing.Foo)
  thingOpt match {
    case Some(Thing.Foo | Thing.Bar) => println("OK")
    case None => println("NOK")
  }
}
```

## problem

Despite `Thing` being unsealed, I still get a warning:

```
match may not be exhaustive.
It would fail on the following input: Some((x: Thing forSome x not in (Bar, Foo)))
```

## expected behavior

I want exhaustivity checking to be disabled for unsealed types regardless of whether the unsealed type is a direct scrutinee of the match or a value nested somewhere inside the scrutinee.

Related discussion: https://contributors.scala-lang.org/t/trouble-with-2-13-4-exhaustivity-checking-being-too-strict/4817

**Comment by dwijnand** [comment]
This is how it's always worked:

```
$ scala -2.13.3
Welcome to Scala 2.13.3 (OpenJDK 64-Bit Server VM, Java 11.0.9).
Type in expressions for evaluation. Or try :help.
final class Thing private();
scala> final class Thing private(); object Thing {
     |   final val Foo = new Thing
     |   final val Bar = new Thing
     | }
class Thing
object Thing

scala> val thingOpt: Option[Thing] = Some(Thing.Foo)
val thingOpt: Option[Thing] = Some(Thing@62158991)

scala>   thingOpt match {
     |     case Some(Thing.Foo | Thing.Bar) => println("OK")
     |     case None => println("NOK")
     |   }
         ^
       warning: match may not be exhaustive.
       It would fail on the following input: Some((x: Thing forSome x not in (Bar, Foo)))
OK
```

**Comment by ghik** [comment]
OK, then why is warning not emitted when `Thing` is the scrutinee (not wrapped in `Option`)?

```
val thing: Thing = Thing.Foo
thing match {
  case Thing.Foo | Thing.Bar => println("OK")
}
```

Whatever the correct behavior should be, currently it seems inconsistent.

**Comment by dwijnand** [comment]
> OK, then why is warning not emitted when `Thing` is the scrutinee (not wrapped in `Option`)?

Because `Thing` isn't sealed, so it doesn't generate a warning unless you have strict unsealed patmat on.

**Comment by ghik** [comment]
@dwijnand
> Because Thing isn't sealed, so it doesn't generate a warning unless you have strict unsealed patmat on.
But it _does_ generate that warning when wrapped into `Option` even though strict unsealed patmat is _off_. It's inconsistent.

**Comment by dwijnand** [comment]
You're right.  But personally I was against the decision to make unsealed types not warn, so I wouldn't want even more warnings turned off by recursing, so I'm anti-motivated to change it... 😅 

But - I'd have to double-check - I think Scala 3 does check recursively (I know it does for the type in a case class, I'm just not sure if it does also for type arguments), which means if someone were to want to submit this change it would probably land.
